Added JobManager to run once a day and create/update configuration jobs.

anthonypos opened the pull request.

Commit:
Added JobManager to run once a day and create/update configuration jobs.

maitlandmarshall commented on MAD.Integration.TableauCRM/Jobs/JobManager.cs:
```diff
@@ -0,0 +1,35 @@
+        public async Task CreateOrUpdateJobs() => await Task.Run(() =>
+        {
+            // Delete inactive jobs
+            foreach (var configuration in this.dbContext.Configuration.Where(y => y.IsActive == false))
```
Instead of doing Task.Run, do var configs = await dbContext.Configuration.Where(..).ToListAsync()

maitlandmarshall commented on MAD.Integration.TableauCRM/Jobs/JobManager.cs:
```diff
@@ -0,0 +1,35 @@
+            }
+
+            // Register active jobs
+            foreach (var configuration in this.dbContext.Configuration.Where(y => y.IsActive))
```
Just do one call to the database. this.dbContext.Configuration.ToListAsync() and then filter locally, as you are pulling everything anyway

maitlandmarshall commented on MAD.Integration.TableauCRM/Jobs/JobManager.cs:
```diff
@@ -0,0 +1,35 @@
+        private readonly IRecurringJobManager recurringJobManager;
+        private readonly IBackgroundJobClient backgroundJobClient;
+
+        public JobManager(ConfigurationDbContext dbContext, IRecurringJobManager recurringJobManager, IBackgroundJobClient backgroundJobClient)
```
Make sure to use the IRecurringJobFactory otherwise the CRON override hooks in settings.json won't work

anthonypos replied:
Doneskies. Wanna add the RemoveIfExists method and any other useful functions to IRecurringJobFactory in the future? Then we don't need IRecurringJobManager.

maitlandmarshall replied:
Do eet. Can you also have a look into how possible it is to implement our own version IRecurringJobManager so we don't have two? Our own implementation should just add the additions of IReccurringJobFactory but under the interface of IReccuringJobManager. It should just act as a proxy for the normal IRecurringJobManager with the extra hooks from IRecurringJobFactory

maitlandmarshall commented on MAD.Integration.TableauCRM/Startup.cs:
```diff
@@ -55,18 +56,8 @@ public async Task PostConfigure(ConfigurationDbContext dbContext, IRecurringJobF
             await dbContext.Database.MigrateAsync();
 
-            // Delete inactive jobs
-            foreach (var configuration in dbContext.Configuration.Where(y => y.IsActive == false))
```
Restore this code otherwise jobs won't be created at start up.

Move the delete / register active jobs code into a service class so it can be reused here and in the JobManager job

Commit:
- Added JobRegistrar service class for updating jobs based on configurations
- Added JobRegistrar to startup and JobManager classes
- Removed Task.Run and replaced with dbContext.Configuration.ToListAsync().

Commit:
- Updated MIFCore.Hangfire to v1.1.0
- Updated unit tests to use new result set factory functionality
- Updated relevant classes to use IRecurringJobManager instead of IRecurringJobFactory
